### PR TITLE
concurrency/mutexmap Move Unlock to after operation

### PR DESCRIPTION
**Pull request by JoshVanL**
# Description

_Please explain the changes you've made_

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #_[issue number]_

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests

### CODE REVIEW

File: concurrency/mutexmap.go [hunk 1]
```diff
@@ -58,50 +58,57 @@ func (a *mutexMap[T]) Lock(key T) {
 	a.lock.RLock()
 	mutex, ok := a.items[key]
 	a.lock.RUnlock()
+	if ok {
```
Comment (artursouza) [inline concurrency/mutexmap.go:61]: Is this just a style change? It seems to do the same as before.

File: concurrency/mutexmap.go [hunk 2]
```diff
@@ -58,50 +58,57 @@ func (a *mutexMap[T]) Lock(key T) {
...
 }
 
 func (a *mutexMap[T]) Unlock(key T) {
 	a.lock.RLock()
 	mutex, ok := a.items[key]
-	a.lock.RUnlock()
 	if ok {
 		mutex.Unlock()
 	}
+	a.lock.RUnlock()
```
Comment (artursouza) [inline concurrency/mutexmap.go:82]: Why is this needed? The lock of the map is now bundled with the lock of the mutex itself. The original design was keeping those separate.

**Review by artursouza — CHANGES_REQUESTED** [review]
Can you provide more context of why this change is needed? The previous design separated the lock of the map with the lock of each mutex.

File: concurrency/mutexmap.go (hunk 2 above)
Comment (JoshVanL) [reply concurrency/mutexmap.go:82]: The change aims to release the outer lock after the inner operation has completed. This prevents race conditions on inner operations of the same key

File: concurrency/mutexmap.go (hunk 1 above)
Comment (artursouza) [reply concurrency/mutexmap.go:61]: OK. I am fine with either way.

File: concurrency/mutexmap.go [hunk 3]
```diff
@@ -58,50 +58,57 @@ func (a *mutexMap[T]) Lock(key T) {
 	a.lock.RLock()
 	mutex, ok := a.items[key]
 	a.lock.RUnlock()
+	if ok {
+		mutex.Lock()
```
Comment (artursouza) [inline concurrency/mutexmap.go:62]: Don't you need to also move this before the map's unlock?

**Review by artursouza — CHANGES_REQUESTED** [review]

File: concurrency/mutexmap.go (hunk 3 above)
Comment (JoshVanL) [reply concurrency/mutexmap.go:62]: No, we have to release the outer lock before locking the inner lock else we will get a deadlock. If the mutex is already locked, other operations like `DeleteUnlock` would never complete.

**Comment by artursouza** [comment]
Can we have a unit test that runs multiple operations in parallel to try a deadlock scenario? I don't disagree that there is a bug but I am concerned about this PR as well for bundling the 2 locks instead of keeping them separate.

File: concurrency/mutexmap.go (hunk 2 above)
Comment (artursouza) [reply concurrency/mutexmap.go:82]: I find it odd that we need to bundle the unlock of item with the unlock of the map.

Comment (artursouza) [reply concurrency/mutexmap.go:82]: Oh. I think I get it now. It is because of the delete.

File: concurrency/mutexmap.go (hunk 3 above)
Comment (elena-kolevska) [reply concurrency/mutexmap.go:62]: I'm not so worried about `DeleteUnlock` because it will never be called before the initial Lock has completed and returned. I think the bigger problem is when we have a delete and a concurrent lock again. Then we would actually get two different mutexes controlling the same resource.

**Comment by JoshVanL** [comment]
Because it is a true race condition of LOC and execution, I'm not sure of a way to write a sensible unit test for it.

**Comment by artursouza** [comment]
It is hard to make a deterministic test for this. Is it possible to make a test that will not cause false positives? Meaning that the race condition will most likely cause the test to fail but not 100% guaranteed. On the other hand, not having the race condition will make the test pass 100%. This way, we can run the test a few times to make sure. It is not ideal but better than visual inspection (aka code review) IMO.

Is there another layer (runtime, maybe) where it can be tested?

**Comment by JoshVanL** [comment]
I can look into doing that. This bug is currently being manifested in Dapr integration tests failing.

**Review by yaron2 — APPROVED** [review]

**Review by artursouza — APPROVED** [review]
Approved under the assumption that integration tests will show the fix.

**Review by artursouza — APPROVED** [review]
